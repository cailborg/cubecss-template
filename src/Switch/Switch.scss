// .toggle {
//   width: 100%;
//   display: flex;
//   align-items: center;
//   justify-content: space-between;
// }

// .switch {
//   margin: 0;
//   padding: 0;
//   border-radius: var(--lp-radii-lg);
//   position: relative;
//   background-color: rgba(var(--lp-colors-neutral-300), 1);
//   cursor: pointer;
//   transition: background-color ease-in-out var(--lp-durations-quick);
//   &.focus-ring {
//     box-shadow: 0 0 0 2px white, 0 0 0 6px dodgerblue;
//   }
// }

// .switch.small {
//   width: 40px;
//   height: 24px;
//   & > span:last-child {
//     width: 24px;
//     border-width: var(--lp-border-widths-thin);
//   }
//   &[aria-checked="true"] > span:last-child {
//     -webkit-transform: translateX(16px);
//     -ms-transform: translateX(16px);
//     transform: translateX(16px);
//   }
// }

// .switch.medium {
//   width: 56px;
//   height: 32px;
//   & > span:last-child {
//     width: 32px;
//   }
//   &[aria-checked="true"] > span:last-child {
//     -webkit-transform: translateX(24px);
//     -ms-transform: translateX(24px);
//     transform: translateX(24px);
//   }
// }

// .switch.large {
//   width: 72px;
//   height: 40px;
//   border-radius: 32px;
//   & > span:last-child {
//     width: 40px;
//   }
//   &[aria-checked="true"] > span:last-child {
//     -webkit-transform: translateX(32px);
//     -ms-transform: translateX(32px);
//     transform: translateX(32px);
//   }
// }

// .switch > * {
//   pointer-events: none;
// }

// .switch:focus {
//   outline-width: 0;
// }

// .slider {
//   display: flex;
//   align-items: center;
//   justify-content: center;
//   position: absolute;
//   top: 0;
//   left: 0;
//   height: 100%;
//   -webkit-transition: var(--lp-durations-quick);
//   transition: var(--lp-durations-quick);
//   border-radius: 50%; 
//   border-width: var(--lp-border-widths-thick); 
//   border-color: rgba(var(--lp-colors-neutral-300), 1);
// }

.disabledSlider {
  @extend .slider;
  background-color: rgba(var(--lp-colors-neutral-400),1);
}

.errorSlider {
  @extend .slider;
  background-color: rgba(var(--lp-colors-error-500), 1);
}

.checkedSlider {
  @extend .slider;
  background-color: white;
  border-color: rgba(var(--lp-colors-primary), 1);
  box-shadow: var(--lp-shadows-low)
}

.uncheckedSlider {
  @extend .slider;
  background-color: rgba(var(--lp-colors-neutral-50), 1);
  border-color: rgba(var(--lp-colors-neutral-300), 1);
}

// .switch[aria-checked="true"] {
//   background-color: rgba(var(--lp-colors-primary), 1);
// }

.toggle label {
  width: calc(100% - 56px - var(--lp-space-xs));
  margin: 0;
  text-align: left;
}