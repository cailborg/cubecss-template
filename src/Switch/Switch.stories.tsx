import { useState } from "react";
import { Meta } from "@storybook/react/types-6-0";
import { Switch } from "./Switch";

export default {
    title: "Components/Switch",
    component: Switch,
} as Meta;

export const Standard = (args: any) => {
    const [checked, setChecked] = useState(false);
    return <Switch {...args} checked={checked} onChange={setChecked} />;
};
Standard.args = { label: "Label Text" };
