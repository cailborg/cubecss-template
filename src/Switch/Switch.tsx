import React from "react";
import "./Switch.scss";
import { FocusRing } from "@react-aria/focus";

export interface SwitchProps {
    checked: boolean;
    onChange: (value: boolean) => void;
    disabled?: boolean;
    label?: string;
    trackingId?: string;
    size: "large" | "medium" | "small";
}

export const Switch: React.FC<SwitchProps> = ({
    checked,
    onChange,
    disabled,
    label,
    trackingId,
    size = "large",
}) => {
    let status;
    switch (true) {
        case disabled:
            status = "disabled";
            break;
        case checked:
            status = "checked";
            break;
        default:
            status = "unchecked";
            break;
    }

    return (
        <div aria-live="polite" aria-atomic="true" className="toggle">
            {label && <label className="font-body text-neutral-800 font-normal">{label}</label>}
            <FocusRing focusRingClass="focus-ring">
                <button
                    type="button"
                    role="switch"
                    aria-checked={!disabled && checked}
                    onClick={() => onChange(!checked)}
                    className={`switch ${size}`}
                    disabled={disabled}
                    data-tracking-id={trackingId}
                >
                    <span className="visually-hidden">{status}</span>
                    <span className={`${status}Slider`}></span>
                </button>
            </FocusRing>
        </div>
    );
};
