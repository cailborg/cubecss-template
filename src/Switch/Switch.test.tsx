import React , { useState }from "react";
import { Switch } from "./Switch";
import renderer from "react-test-renderer";

it("renders correctly", () => {
  const [checked, setChecked] = useState(false);
  const tree = renderer.create(<Switch checked={checked} size="large" onChange={setChecked}></Switch>).toJSON();
  expect(tree).toMatchSnapshot();
});
