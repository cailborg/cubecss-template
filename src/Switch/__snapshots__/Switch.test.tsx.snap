// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="css-1povu0j"
>
  <div
    className="css-1ov7xgf"
    size="medium"
  >
    <p
      className="css-d21g04-Text"
    >
      someText
    </p>
  </div>
  <label
    className="css-1y0vz5m"
    size="medium"
  >
    <input
      className=""
      onBlur={[Function]}
      onFocus={[Function]}
      type="checkbox"
    />
    <span />
  </label>
</div>
`;
